hphp/tools/tc-print/offline-code.cpp

```diff
@@ -124,7 +121,326 @@ TCA OfflineX86Code::getTransJmpTargets(const TransRec *transRec,
+                            BCMappingInfo bcMappingInfo,
+                            bool printAddr,
+                            bool printBinary) {
+
+  if (codeLen == 0) return;
+ 
+  char codeStr[MAX_INSTR_ASM_LEN];
+  Instruction* code = (Instruction*) alloca(codeLen);
+  Instruction* frontier;
+  TCA ip;
+  size_t	currBC = 0;
+
+  auto const offset = codeStartAddr - fileStartAddr;
+  if (fseek(file, offset, SEEK_SET)) {
+    error("disasm error: seeking file");
+  }
+
+  size_t readLen = fread(code, codeLen, 1, file);
+  if (readLen != 1) {
+    error("Failed to read {} bytes at offset {} from code file due to {}",
+          codeLen, offset, feof(file) ? "EOF" : "read error");
+  }
+
+  Decoder dec;
+  Disassembler dis(codeStr, MAX_INSTR_ASM_LEN);
+  dis.MapCodeAddress((int64_t)codeStartAddr, code);
+  dec.AppendVisitor(&dis);
+  int64_t callAddr = 0;
+  Instr insn;
+
+  const Instr MOVZ_x18 = MOVZ_x | 0x12;            // movz x18, #0x...
+  const Instr MOVK_x18 = MOVK_x | 0x12;            // movk x18, #0x... {, lsl #N}
+  const Instr MoveWideImmX18Mask = 0xFF800012;
+
+  const Instr LDR_x_litx18 = LDR_x_lit | 0xffffd2; // ldr  x18, pc-8
+  const Instr BLR_x18 = BLR | 0x240;               // blr  x18
+
+  for (frontier = code, ip = codeStartAddr; frontier < code + codeLen; ) {
+    dec.Decode(frontier);
+
+    currBC = printBCMapping(bcMappingInfo, currBC, (TCA)ip);
+    if (printAddr) printf("%14p: ", ip);
+
+    if (printBinary) {
+      printf("%08" PRIx32 , *reinterpret_cast<int32_t *>(frontier));
+      printf("%10s","");
+    }
+
+    // Shadow potential call destinations based on fixed sequence.
+    // This needs to match code generation in JIT.
+    //
+    //   movz x18, #...        ldr x18, pc-8
+    //   movk x18, #...        blr x18
+    //   blr x18
+
+    insn = frontier->InstructionBits();
+    if ((insn & MoveWideImmX18Mask) == MOVZ_x18) {
+      callAddr = static_cast<uint32_t>((insn & 0x1fffd0) >> ImmMoveWide_offset);
+
+    } else if ((insn & MoveWideImmX18Mask) == MOVK_x18) {
+      callAddr |= ((insn & 0x1fffe0) >> 5) << (16 * ((insn & 0x600000) >> ShiftMoveWide_offset));
+
+    } else if (insn == LDR_x_litx18) {
+      callAddr = static_cast<int64_t>((frontier-4)->InstructionBits()) << 32;
+      callAddr |= (frontier-8)->InstructionBits();
+      callAddr = callAddr + (int64_t)ip;
+      callAddr = callAddr + (int64_t)kInstructionSize;
+    }
+        
+    string callDest="";
+    if ((insn & BLR_x18) == BLR_x18) {
+      callDest = getSymbolName((TCA)callAddr);
+      callAddr = 0;
+    }
+
+    if (! perfEvents.empty()) {
+      printEventStats((TCA)ip, kInstructionSize, perfEvents);
+    } else {
+      printf("%48s", "");
+    }
+
+    printf("%s%s\n", codeStr, callDest.c_str());
+
+    frontier += kInstructionSize;	
+    ip += kInstructionSize;
+  }
+}
+
+#elif defined(__x86_64__)
+#if defined(HAVE_LIBXED)
+
+// x86 below here
+// TODO: move to separate file
```

Comment:
I agree that putting the target-specific code on separate files would be better. Could you please do so?